Sources/NIO/ByteBuffer-core.swift

```diff
@@ -800,6 +804,69 @@ extension ByteBuffer {
     }
 }
 
+extension ByteBuffer {
+    /// Copies `length` `bytes` starting at the `fromIndex` to `toIndex`. Does not move the writer index.
+    ///
+    /// - Note: Overlapping ranges, for example `copyBytes(at: 1, to: 2, length: 5)` are allowed.
+    /// - Precondition: The range represented by `fromIndex` and `length` must be readable bytes,
+    ///     that is: `fromIndex >= readerIndex` and `fromIndex + length <= writerIndex`.
+    /// - Parameter fromIndex: The index of the first byte to copy.
+    /// - Parameter toIndex: The index into to which the first byte will be copied.
+    /// - Parameter length: The number of bytes which should be copied.
+    @discardableResult
+    @inlinable
+    public mutating func copyBytes(at fromIndex: Int, to toIndex: Int, length: Int) throws -> Int {
+        switch length {
+        case ..<0:
+            throw CopyBytesError.negativeLength
+        case 0:
+            return 0
+        default:
+            ()
+        }
+        guard self.readerIndex <= fromIndex && fromIndex + length <= self.writerIndex else {
+            throw CopyBytesError.unreadableSourceBytes
+        }
+
+        if !isKnownUniquelyReferenced(&self._storage) {
+            let newEndIndex = max(self._writerIndex, _toIndex(toIndex + length))
+            self._copyStorageAndRebase(capacity: newEndIndex)
+        }
+
+        self._ensureAvailableCapacity(_Capacity(length), at: _toIndex(toIndex))
+        self.withVeryUnsafeBytes { ptr in
+            let srcPtr = UnsafeRawBufferPointer(start: ptr.baseAddress!.advanced(by: fromIndex), count: length)
+            self._setBytesAssumingUniqueBufferAccess(srcPtr, at: _toIndex(toIndex))
+        }
+
+        return length
+    }
+
+    /// Errors thrown when calling `copyBytes`.
+    public struct CopyBytesError: Error {
+        internal enum BaseError: Error, Hashable {
```

Comment:
This internal enum doesn't need to conform to `Error`. It may be a good idea to prefix this with an underscore as well to discourage its use internally.

Reply:
Made it `private` instead.